scipy/stats/stats.py

```diff
@@ -4153,6 +4154,178 @@ def binary_search(n, n1, n2, side):
+    >>> import scipy.stats as stats
+    >>> TX, pvalue = stats.barnard_exact([[7, 12], [8, 3]])
+    >>> pvalue
+    0.0681...
+
+    """
+    # ipdb.set_trace()
+    if nuisance_nbre < 0:
+        raise ValueError(
+            f"Number of nuisance's samples, {nuisance_nbre}, must be non-negative."
+            )
+
+    c = np.asarray(
+        table,
+        dtype=np.int64)
+
+    if not c.shape == (2, 2):
+        raise ValueError("The input `table` must be of shape (2, 2).")
+
+    if np.any(c < 0):
+        raise ValueError("All values in `table` must be nonnegative.")
+
+    if 0 in c.sum(axis=0) or 0 in c.sum(axis=1):
+        # If both values in a row or column are zero, the p-value is 1 and
+        # the odds ratio is NaN.
+        return np.nan, 1.0
+
+    total_c1, total_c2 = c.sum(axis=0)
+    n = total_c1 + total_c2
+
+    x1 = np.arange(total_c1 + 1, dtype=np.int64).reshape(-1, 1)
+    x2 = np.arange(total_c2 + 1, dtype=np.int64).reshape(1, -1)
+
+    def compute_score_statistic(c1, c2, x1, x2, variance):
+        p1, p2 = x1 / c1, x2 / c2
+
+        if variance == "pooled":
+            p = (x1 + x2) / (c1 + c2)
+            var_p1_p2 = p * (1 - p) * (1 / c1 + 1 / c2)
+        elif variance == "unpooled":
+            var_p1_p2 = p1 * (1 - p1) / c1 + p2 * (1 - p2) / c2
+        else:
+            msg = (
+                f"`variance` should be one of {'pooled', 'unpooled'}.\
+                value: {variance}"
+            )
+            raise ValueError(msg)
+
+        # To avoid warning when dividing by 0
+        np.seterr(divide='ignore', invalid='ignore')
+        score = (p2 - p1) / np.sqrt(var_p1_p2)
+        np.seterr(divide=None, invalid=None)
+
+        return score
+
+    TX = compute_score_statistic(
+        total_c1,
+        total_c2,
+        x1,
+        x2,
+        variance)
+    if alternative == "two-sided":
+        TX_obs = TX[
+            c[0, 0], c[0, 1]
+        ]
+        TX = np.abs(TX)
+        idx = TX >= abs(TX_obs)
+    elif alternative == "less":
+        TX_obs = TX[
+            c[0, 0], c[0, 1]
+        ]
+        idx = TX <= TX_obs
+
+    elif alternative == "greater":
+        TX_obs = TX[
+            c[0, 0], c[0, 1]
+        ]
+        idx = TX >= TX_obs
+    else:
+        msg = f"`alternative` should be one of {'two-sided', 'less', 'greater'}. value: {alternative}"
+        raise ValueError(msg)
+
+    # Pass x1 and x2 in dimension 3
+    x1 = x1.reshape(-1, 1, 1)
+    x2 = x2.reshape(1, -1, 1)
+
+    nuisance_arr = np.linspace(
+        start=0, stop=1, num=nuisance_nbre + 2)[1:-1]
+    nuisance_arr = nuisance_arr[1:-1]  # Remove 0 and 1
+    # Reshape in dimension 3 array
+    nuisance_arr = nuisance_arr.reshape(1, 1, -1)
+
+    def compute_gammaln_combination(n):
+        # Compute all combination of C(n, k)
+        gammaln_arr = special.gammaln(np.arange(n + 1) + 1)
+        return special.gammaln(n + 1) - gammaln_arr - gammaln_arr[::-1]
```

Comment:
This function should be defined outside the mail function. Maybe `_barnard…_compute…`